pkg/controller/controller_utils.go

```diff
@@ -357,6 +357,48 @@ func ClusterAPIMachineProviderConfigFromMachineSetSpec(machineSetSpec *clusterop
 	}, nil
 }
 
+// AWSMachineProviderStatusFromClusterAPIMachine gets the cluster-operator MachineSetSpec from the
+// specified cluster-api MachineSet.
+func AWSMachineProviderStatusFromClusterAPIMachine(m *clusterapi.Machine) (*clusteroperator.AWSMachineProviderStatus, error) {
+	if m.Status.ProviderStatus == nil {
+		return &clusteroperator.AWSMachineProviderStatus{}, nil
+	}
+	obj, gvk, err := api.Codecs.UniversalDecoder(clusteroperator.SchemeGroupVersion).Decode([]byte(m.Status.ProviderStatus.Raw), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	status, ok := obj.(*clusteroperator.AWSMachineProviderStatus)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected object: %#v", gvk)
+	}
+	return status, nil
+}
+
+// ClusterAPIMachineProviderStatusFromAWSMachineProviderStatus gets the cluster-api ProviderConfig for a Machine template
+// to store the cluster-operator MachineSetSpec.
+func ClusterAPIMachineProviderStatusFromAWSMachineProviderStatus(awsStatus *clusteroperator.AWSMachineProviderStatus) (*runtime.RawExtension, error) {
+	awsStatus.TypeMeta = metav1.TypeMeta{
+		APIVersion: clusteroperator.SchemeGroupVersion.String(),
+		Kind:       "AWSMachineProviderStatus",
+	}
+	serializer := jsonserializer.NewSerializer(jsonserializer.DefaultMetaFactory, api.Scheme, api.Scheme, false)
+	var buffer bytes.Buffer
+	err := serializer.Encode(awsStatus, &buffer)
+	if err != nil {
+		return nil, err
+	}
+	// Encoding json includes a trailing line feed, which will not be present on objects
+	// read back from etcd. If we see the last byte as a line feed, remove it, allowing us
+	// to accurately compare new/old status.
+	rawBytes := buffer.Bytes()
+	if rawBytes[len(rawBytes)-1] == 10 {
+		rawBytes = rawBytes[:len(rawBytes)-1]
+	}
```

Comment:
PR updated, added a hack here, status was always showing a change because the freshly encoded raw extension gets a trailing newline from the json encoder, whereas this is stripped off when written to and read back from etcd.

Reply:
Just found bytes.TrimSpace, replacing.